File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/infrastructure/GiftCardRepositoryTest.java
package it.giannibombelli.wsc2026.giftcard.infrastructure;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Verifica il contratto del repository senza ripetere le regole di business di dominio.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class GiftCardRepositoryTest {
    private GiftCardRepository repository;

    @BeforeAll
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeFileDb("giftcard", getClass().getSimpleName());
        repository = new SqliteGiftCardRepository(dataSource);
    }

    @Nested
    class Save {
        @Test
        void shouldPersistNewCard() {
            GiftCardId cardId = EntityId.generate(GiftCardId::new);
            GiftCard original = GiftCard.issue(cardId);

            repository.save(original);
            Optional<GiftCard> reloaded = repository.findById(cardId);

            assertThat(reloaded).isPresent();
            GiftCard found = reloaded.get();
            assertThat(found.id()).isEqualTo(cardId);
            assertThat(found.balance()).isEqualTo(Money.zero());
        }

        @Test
        void shouldUpdateBalance() {
            GiftCardId cardId = EntityId.generate(GiftCardId::new);
            GiftCard credited = new GiftCard(cardId, new Money(new BigDecimal("37.75")));

            repository.save(credited);
            Optional<GiftCard> reloaded = repository.findById(cardId);

            assertThat(reloaded).isPresent();
            assertThat(reloaded.get().balance()).isEqualTo(new Money(new BigDecimal("37.75")));
        }
    }

    @Nested
    class FindById {
        @Test
        void shouldReturnEmptyWhenNotFound() {
            GiftCardId nonExistentId = EntityId.generate(GiftCardId::new);

            Optional<GiftCard> reloaded = repository.findById(nonExistentId);

            assertThat(reloaded).isEmpty();
        }
    }
}

File: typescript/tests/giftcard/infrastructure/giftCardRepository.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { GiftCard } from '@/giftcard/domain/giftcard/giftCard.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import type { GiftCardRepository } from '@/giftcard/domain/ports/giftCardRepository.js';
import { SqliteGiftCardRepository } from '@/giftcard/infrastructure/sqliteGiftCardRepository.js';
import { DatabaseSetup } from '../../testsupport/databaseSetup.js';

describe('GiftCardRepository', () => {
  let repository: GiftCardRepository;

  beforeAll(() => {
    const database = DatabaseSetup.initializeFileDb('giftcard', 'GiftCardRepositoryTest');
    repository = new SqliteGiftCardRepository(database);
  });

  describe('save', () => {
    it('should persist new card', () => {
      const cardId = generateId((value) => new GiftCardId(value));
      const original = GiftCard.issue(cardId);

      repository.save(original);
      const reloaded = repository.findById(cardId);

      expect(reloaded).not.toBeNull();
      expect(reloaded!.id()).toEqual(cardId);
      expect(reloaded!.balance()).toEqual(Money.zero());
    });

    it('should update balance', () => {
      const cardId = generateId((value) => new GiftCardId(value));
      const credited = new GiftCard(cardId, new Money(37.75));

      repository.save(credited);
      const reloaded = repository.findById(cardId);

      expect(reloaded).not.toBeNull();
      expect(reloaded!.balance()).toEqual(new Money(37.75));
    });
  });

  describe('findById', () => {
    it('should return empty when not found', () => {
      const nonExistentId = generateId((value) => new GiftCardId(value));

      const reloaded = repository.findById(nonExistentId);

      expect(reloaded).toBeNull();
    });
  });
});
